### PR TITLE
Enable percpu_arena:percpu for jemalloc

### PR DESCRIPTION
**Pull request by azat**
**Conflicts with: #11086**

Let's see on performance test results.

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Enable percpu_arena:percpu for jemalloc (This will reduce memory fragmentation due to thread pool)

Refs: #10818

### CODE REVIEW

**Comment by qoega** [comment]
@akuzm you can check memory diff will be found in performance test for this PR

**Comment by akuzm** [comment]
> @akuzm you can check memory diff will be found in performance test for this PR

This PR just changes where the memory is allocated, but won't change per-query memory usage.

**Comment by akuzm** [comment]
What we could see in performance tests is probably the reduction in global memory usage -- metric_log is exported, but I'm not sure which metric we need.

**Comment by akuzm** [comment]
`MemoryStatisticsOS` could be useful, but it's in `system.asynchronous_metrics`, and there is no log for them yet https://github.com/ClickHouse/ClickHouse/issues/9429

**Comment by azat** [comment]
@akuzm `MemoryTracking` from `system.metrics` periodically resets to the RSS, but I doubt that this is enough or even useful

https://github.com/ClickHouse/ClickHouse/blob/f98ffdbc4ca7ff32350c4b2fd5f672d70168cb2d/src/Interpreters/AsynchronousMetrics.cpp#L155

Maybe there is some internal monitoring for this machines?

>This PR just changes where the memory is allocated, but won't change per-query memory usage.

And (which is more important) when it will be freed.
But as a side effect it can add [latency](https://github.com/jemalloc/jemalloc/issues/713), which can make server slower

**Comment by alexey-milovidov** [comment]
Both PRs are Ok according to performance tests.
The question is - what is the actual difference in memory fragmentation that you are experiencing.

**Review by alexey-milovidov — APPROVED** [review]
I'm ready to merge it if there is any evidence of lower memory usage.

**Comment by azat** [comment]
Using latest upstream/master (19067621190677a0589ae3af84514f9a8cd94356)

And started server with `max_server_memory_usage=8G`:

```
clickhouse-server -- --max_server_memory_usage=$((8<<30))
```

Then I get the test from #10818

```
clickhouse-benchmark --max_threads 16 <<< "select WatchID from hits_100m_obfuscated group by WatchID limit 10"
```

And got the following:

| type                                         | fail after   | VIRT (GiB)   | RSS (GiB) (WAS -> after 1min)   | nTH               | 
| -------------------------------------------- | ------------ | ------------ | ------------------------------- | ----------------- | 
| basic                                        | 19           | 19.8         | 9.2                             | 88                | 
| prefill                                      | 19           | 52.6         | 30                              | 1K+               | 
| background_thread                            | 15           | 20           | 8.4 -> 0.3                      | 116 (+jemalloc)   | 
| background_thread + prefill                  | 24           | 53           | 13 -> 0.4                       | 1K+               | 
| percpu_narena                                | 22           | 18           | 8 -> 5                          | 116 (+jemalloc)   | 
| percpu_narena + prefill                      | 15           | 25           | 7.5 -> 5                        | 1K+               | 
| background_thread + percpu_narena            | 42           | 15           | 7.5 -> 0.3                      | 116 (+jemalloc)   | 
| background_thread + percpu_narena + prefill  | 15           | 23           | 8 -> 0.7                        | 1K+ (+jemalloc)   |

Notes:
- `fail after` "difference" should not be considered as anything
- `fail after` queries fails due to memory tracking issues
- `prefill` - ThreadPool was pre-filled with other threads to increase number of active threads (with the following query `select count(*) from (select * from numbers_mt(1000000) settings max_threads=1000)`)

So basically `percpu_narena` shows itself better then everything other (no need to enable `background_thread` I guess), but my main concern about affecting mixed payload (performance tests are a little bit synthetic I guess), but maybe it will works well

**Comment by alexey-milovidov** [comment]
> but my main concern about affecting mixed payload (performance tests are a little bit synthetic I guess), but maybe it will works well

Let's be optimistic.

**Comment by alexey-milovidov** [comment]
Not ready.

clickhouse-benchmark prints this message at startup:
```
<jemalloc>: Conf string ends with key
```

**Comment by alexey-milovidov** [comment]
And server RSS is 35 GiB, so the issue #10818 is not solved.
